### PR TITLE
Effect crasher fix 

### PR DESCRIPTION
**Pull request by daschuer**
This PR fixes a crasher due to unsafe handling of the effect state lifetime and concurrent access to m_chainStatusForChannelMatrix in EngineEffectChain::deleteStatesForInputChannel() 
https://bugs.launchpad.net/mixxx/+bug/1775497

This has been introduced in Mixxx 2.1 via #1468.
Under normal conditions the crash does not happen, but it happens whenever the order of execution of the main and engine thread is disturbed due to an overload condition. A locking that is able to prevent this is missing. 
The delete attempt from in the main thread might be delayed and the engine is still accessing the memory. This causes a crash. The other issue is that the m_chainStatusForChannelMatrix is written concurrently from the engine and the master thread which must not happen. It must only change state at the beginning of the new engine callback. #4487 

This PR also fixes https://bugs.launchpad.net/mixxx/+bug/1966607 where the effect was not ramped up/down during loading a new track. 

To solve both issues, enabling and disabling of an effect is no done in the same way regardless of which control disables the effect. The effect state is initialized on demand, but it is keep until the effect is unloaded under all conditions.  
Before it was kept if the effect was disabled via the effect enable button, but it was deleted when the effect was disabled via a routing button.  

This allows finally to delete all the brittle code around EngineEffectChain::deleteStatesForInputChannel(). 

This is big PR and the crasher happens only rarely, I have decided to build this PR on top of 2.4, which allows us to test it well during beta phase. A backport to 2.3 is not simple, because of the recent effect refactoring. 

I have tried to build self contained commits, that can be reviewed if the whole PR

### CODE REVIEW

**Comment by daschuer** [comment]
This one merges cleanly again.

**Comment by ronso0** [comment]
> The following tests FAILED:
	288 - EngineSyncTest.EjectTrackSyncRemains (SEGFAULT)
	363 - HotcueControlTest.SavedLoopUnloadTrackWhileActive (SEGFAULT)
	530 - AdjustReplayGainTest.AdjustReplayGainUpdatesPregain (SEGFAULT)
Errors while running CTest

Who is able to review this?

**Comment by daschuer** [comment]
This PR has discovered a missing null check introduced here: 
https://github.com/mixxxdj/mixxx/pull/4771

File: src/engine/effects/engineeffect.cpp [hunk 1]
```diff
@@ -46,24 +46,16 @@ EngineEffect::~EngineEffect() {
...
-    if (kEffectDebugOutput) {
-        qDebug() << "EngineEffect::loadStatesForInputChannel" << this
-                 << "loading states for input" << inputChannel;
-    }
-    m_pProcessor->loadStatesForInputChannel(inputChannel, pStatesMap);
-}
 
-void EngineEffect::deleteStatesForInputChannel(ChannelHandle inputChannel) {
-    m_pProcessor->deleteStatesForInputChannel(inputChannel);
+    //TODO: get actual configuration of engine
```
Comment (ywwg) [inline src/engine/effects/engineeffect.cpp:54]: I think this TODO needs to be addressed?

**Review by ywwg — CHANGES_REQUESTED** [review]
This looks great, much cleaner!  I agree that we can merge to 2.4 and do some testing and fixing.  I made a couple of notes.

**Comment by daschuer** [comment]
Done.

**Review by ywwg — APPROVED** [review]

**Comment by ywwg** [comment]
ok to merge

**Pull request merged**